File: app/src/main/java/com/yibao/music/activity/SearchLyricsActivity.java
package com.yibao.music.activity;

import android.content.Intent;
import android.graphics.drawable.AnimationDrawable;
import android.view.View;

import androidx.viewpager2.widget.ViewPager2;

import com.yibao.music.R;
import com.yibao.music.adapter.SearchLyricsPagerAdapter;
import com.yibao.music.base.BaseObserver;
import com.yibao.music.base.bindings.BaseBindingActivity;
import com.yibao.music.databinding.ActivitySearchLyricsBinding;
import com.yibao.music.model.qq.SearchLyricsBean;
import com.yibao.music.model.qq.SongLrc;
import com.yibao.music.network.RetrofitHelper;
import com.yibao.music.util.Constant;
import com.yibao.music.util.LogUtil;
import com.yibao.music.util.NetworkUtil;
import com.yibao.music.util.ToastUtil;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

/**
 * @author lsp
 * createDate：2019/12/26 0026 14:54
 * className   SearchLyricsActivity
 * Des：TODO
 */
public class SearchLyricsActivity extends BaseBindingActivity<ActivitySearchLyricsBinding> {
    protected final String TAG = "====" + this.getClass().getSimpleName() + "    ";

    private List<SearchLyricsBean> mLyricsBeanList;
    private String mSongMid;


    @Override
    public void initView() {

    }

    @Override
    public void initData() {
        mLyricsBeanList = new ArrayList<>();
        String mSongName = getIntent().getStringExtra(Constant.SONG_NAME);
        String mSongArtist = getIntent().getStringExtra(Constant.SONG_ARTIST);
        LogUtil.d(TAG, mSongName + " == " + mSongArtist);
        if (mSongName != null && mSongArtist != null) {
            mBinding.editSearchLyricsName.setText(mSongName);
            mBinding.editSearchLyricsArtist.setText(mSongArtist);
        }
        searchLyrics(false);

    }


    @Override
    public void initListener() {
        mBinding.ivSearchDown.setOnClickListener(v -> finish());
        mBinding.ivSearchLyrics.setOnClickListener(v -> searchLyrics(mBinding.editSearchLyricsName.getText().toString().trim().isEmpty()));
        mBinding.tvSearchLyricsComplete.setOnClickListener(v -> searchComplete());
        mBinding.vp2SearchLyrics.registerOnPageChangeCallback(new ViewPager2.OnPageChangeCallback() {
            @Override
            public void onPageSelected(int position) {
                super.onPageSelected(position);
                SearchLyricsBean searchLyricsBean = mLyricsBeanList.get(position);
                mSongMid = searchLyricsBean.getSongMid();
                LogUtil.d(TAG, mSongMid);
                mBinding.tvLyricsPageIndex.setText(String.valueOf(position + 1));
            }
        });
    }


    private void searchLyrics(boolean isNeedArtist) {
        showProgress();
        mLyricsBeanList.clear();
        if (NetworkUtil.isNetworkConnected()) {
            String songName = mBinding.editSearchLyricsName.getText().toString().trim();
            String singer = mBinding.editSearchLyricsArtist.getText().toString().trim();
            RetrofitHelper.getMusicService().getLrc(songName).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).subscribe(new BaseObserver<SongLrc>() {
                @Override
                public void onNext(SongLrc songLrc) {
                    List<SongLrc.DataBean.LyricBean.ListBean> list = songLrc.getData().getLyric().getList();
                    for (SongLrc.DataBean.LyricBean.ListBean listBean : list) {
                        String content = listBean.getContent();
                        String songSinger = listBean.getSinger().get(0).getName();
                        String songNames = listBean.getSongname();
                        if (isNeedArtist) {
                            if (!Constant.NO_LYRICS.equals(content) && !Constant.PURE_MUSIC.equals(content) && songName.equals(songNames) && songSinger.contains(singer)) {
                                SearchLyricsBean lyricsBean = new SearchLyricsBean(listBean.getSongmid(), listBean.getContent());
                                if (!mLyricsBeanList.contains(lyricsBean)) {
                                    mLyricsBeanList.add(lyricsBean);
                                }
                            }

                        } else {
                            if (!Constant.NO_LYRICS.equals(content) && !Constant.PURE_MUSIC.equals(content) && songName.equals(songNames)) {

                                SearchLyricsBean lyricsBean = new SearchLyricsBean(listBean.getSongmid(), listBean.getContent());
                                if (!mLyricsBeanList.contains(lyricsBean)) {
                                    mLyricsBeanList.add(lyricsBean);
                                }
                            }
                        }
                    }
                    mBinding.ivSearchLyricsLoading.setVisibility(View.GONE);
                    setTvIndex(mLyricsBeanList.size());
                    SearchLyricsPagerAdapter pagerAdapter2 = new SearchLyricsPagerAdapter(SearchLyricsActivity.this, mLyricsBeanList);
                    mBinding.vp2SearchLyrics.setAdapter(pagerAdapter2);

                }

                @Override
                public void onError(Throwable e) {
                    LogUtil.d(TAG, "歌词搜索加载错误 " + e.getMessage());
                }
            });
        } else {
            mBinding.ivSearchLyricsLoading.setVisibility(View.GONE);
            ToastUtil.show(this, Constant.NO_NETWORK);
        }

    }

    private void showProgress() {
        mBinding.ivSearchLyricsLoading.setVisibility(View.VISIBLE);
        AnimationDrawable animation = (AnimationDrawable) mBinding.ivSearchLyricsLoading.getBackground();
        animation.start();
    }

    private void setTvIndex(int size) {
        String lyricsCount = size > 1 ? "搜索到" + size + "个结果 (右滑查看多个歌词)" : "搜索到" + size + "个结果";
//        String lyricsCount = "搜索到" + size + "个结果";
        mBinding.tvSearchLyricsCount.setText(lyricsCount);
        if (size != 0) {
            mBinding.tvLyricsPageIndex.setText("1");
        }
    }

    private void searchComplete() {
        LogUtil.d(TAG, "搜索歌词完成");
        Intent intent = new Intent();
        intent.putExtra(Constant.SONGMID, mSongMid);
        setResult(Constant.SELECT_LYRICS, intent);
        finish();
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        finish();
    }

    @Override
    public void finish() {
        super.finish();
        overridePendingTransition(0, R.anim.dialog_push_out);
    }


}

File: app/src/main/java/com/yibao/music/activity/SearchLyricsActivity.kt
package com.yibao.music.activity

import android.content.Intent
import android.graphics.drawable.AnimationDrawable
import android.view.View
import androidx.viewpager2.widget.ViewPager2.OnPageChangeCallback
import com.yibao.music.R
import com.yibao.music.adapter.SearchLyricsPagerAdapter
import com.yibao.music.base.bindings.BaseBindingActivity
import com.yibao.music.databinding.ActivitySearchLyricsBinding
import com.yibao.music.model.qq.SearchLyricsBean
import com.yibao.music.util.Constant
import com.yibao.music.util.LogUtil
import com.yibao.music.viewmodel.SearchViewModel

/**
 * @author lsp
 * createDate：2019/12/26 0026 14:54
 * className   SearchLyricsActivity
 * Des：TODO
 */
class SearchLyricsActivity : BaseBindingActivity<ActivitySearchLyricsBinding>() {


    private val searchViewModel = SearchViewModel()

    private val mLyricsBeanList: MutableList<SearchLyricsBean> = ArrayList()
    private var mSongMid: String? = null

    override fun initView() {
    }

    override fun initData() {
        val songName = intent.getStringExtra(Constant.SONG_NAME)
        val singer = intent.getStringExtra(Constant.SONG_ARTIST)
        LogUtil.d(TAG, "$songName == $singer")
        if (songName != null && singer != null) {
            mBinding.editSearchLyricsName.setText(songName)
            mBinding.editSearchLyricsArtist.setText(singer)
            searchViewModel.searchLyrics(songName, singer, false)
        }
    }


    override fun initListener() {
        mBinding.ivSearchDown.setOnClickListener { v: View? -> finish() }
        mBinding.ivSearchLyrics.setOnClickListener { v: View? ->
            showProgress()
            val isNeed =
                mBinding.editSearchLyricsName.text.toString().trim { it <= ' ' }.isEmpty()
            val songName = mBinding.editSearchLyricsName.text.toString().trim { it <= ' ' }
            val singer = mBinding.editSearchLyricsArtist.text.toString().trim { it <= ' ' }
            searchViewModel.searchLyrics(songName, singer, isNeed)
        }
        mBinding.tvSearchLyricsComplete.setOnClickListener { v: View? -> searchComplete() }
        mBinding.vp2SearchLyrics.registerOnPageChangeCallback(object : OnPageChangeCallback() {
            override fun onPageSelected(position: Int) {
                super.onPageSelected(position)
                val searchLyricsBean = mLyricsBeanList[position]
                mSongMid = searchLyricsBean.songMid
                LogUtil.d(TAG, mSongMid)
                val lyricsIndex = position + 1
                mBinding.tvLyricsPageIndex.text = lyricsIndex.toString()
            }
        })
    }

    override fun onResume() {
        super.onResume()
        searchViewModel.lrcViewModel.observe(this) { lyricsList: List<SearchLyricsBean>? ->
            val animation = mBinding.ivSearchLyricsLoading.background as AnimationDrawable
            animation.stop()
            if (!lyricsList.isNullOrEmpty()) {
                mLyricsBeanList.clear()
                mLyricsBeanList.addAll(lyricsList)
                mBinding.ivSearchLyricsLoading.visibility = View.GONE
                setTvIndex(lyricsList.size)
                val pagerAdapter2 = SearchLyricsPagerAdapter(this@SearchLyricsActivity, lyricsList)
                mBinding.vp2SearchLyrics.adapter = pagerAdapter2
            } else {
                mBinding.tvSearchLyricsCount.text = "没搜索到歌词"
            }
        }
    }

    private fun showProgress() {
        mBinding.ivSearchLyricsLoading.visibility = View.VISIBLE
        val animation = mBinding.ivSearchLyricsLoading.background as AnimationDrawable
        animation.start()
    }

    private fun setTvIndex(size: Int) {
        val lyricsCount =
            if (size > 1) "搜索到" + size + "个结果 (左右滑动查看多个歌词)" else "搜索到" + size + "个结果"
        mBinding.tvSearchLyricsCount.text = lyricsCount
        if (size != 0) {
            mBinding.tvLyricsPageIndex.text = "1"
        }
    }

    private fun searchComplete() {
        LogUtil.d(TAG, "搜索歌词完成")
        // 传递歌曲ID ，用于搜索歌词。
        val intent = Intent()
        intent.putExtra(Constant.SONGMID, mSongMid)
        setResult(Constant.SELECT_LYRICS, intent)
        finish()
    }


    override fun finish() {
        super.finish()
        overridePendingTransition(0, R.anim.dialog_push_out)
    }
}
